python/mxnet/gluon/rnn/rnn_cell.py

```diff
@@ -719,6 +719,58 @@ def unroll(self, length, inputs, begin_state=None, layout='NTC', merge_outputs=N
         return outputs, states
 
 
+class VariationalDropoutCell(ModifierCell):
+    """
+    Applies Variational Dropout on base cell.
+    (https://arxiv.org/pdf/1512.05287.pdf).
+    """
+    def __init__(self, base_cell, vardrop_outputs=0., vardrop_states=0.):
+        assert not isinstance(base_cell, BidirectionalCell), \
+            "BidirectionalCell doesn't support vardrop since it doesn't support step. " \
+            "Please add VariationalDropoutCell to the cells underneath instead."
+        assert not isinstance(base_cell, SequentialRNNCell) or not base_cell._bidirectional, \
+            "Bidirectional SequentialRNNCell doesn't support vardrop. " \
+            "Please add VariationalDropoutCell to the cells underneath instead."
+        super(VariationalDropoutCell, self).__init__(base_cell)
+        self.vardrop_outputs = vardrop_outputs
+        self.vardrop_states = vardrop_states
+        self.vardrop_outputs_mask = None
+        self.vardrop_states_mask = None
+
+    def _alias(self):
+        return 'vardrop'
+
+    def reset(self):
+        super(VariationalDropoutCell, self).reset()
+        self.vardrop_outputs_mask = None
+        self.vardrop_states_mask = None
+
+    def hybrid_forward(self, F, inputs, states):
+        cell = self.base_cell
+
+        if self.vardrop_states:
+            if self.vardrop_states_mask is None:
```

Comment:
Correct me if I'm wrong.
I think in https://arxiv.org/pdf/1512.05287.pdf. Variational drop was droping input and hidden instead of memory cell. Droping cell was from  RnnDrop: A Novel Dropout for RNNs in ASR. But I guess either way give similar results.